File: app/src/main/java/com/zest/android/detail/DetailPresenter.java
package com.zest.android.detail;

import android.text.TextUtils;
import android.util.Log;

import com.zest.android.data.Recipe;
import com.zest.android.data.source.DetailRepository;

import org.greenrobot.greendao.annotation.NotNull;

import java.util.Arrays;
import java.util.List;

/**
 * Created by ZARA on 9/11/2017.
 */

public class DetailPresenter implements DetailContract.UserActionsListener {

    private static final String TAG = DetailPresenter.class.getSimpleName();
    private final DetailContract.View mDetailView;
    private final DetailRepository mDetailRepository;


    public DetailPresenter(@NotNull DetailContract.View detailView,
                           @NotNull DetailRepository detailRepository) {
        this.mDetailView = detailView;
        this.mDetailRepository = detailRepository;
        mDetailView.setPresenter(this);
    }

    @Override
    public void start() {

    }


    @Override
    public List<String> loadTagTitles(Recipe recipe) {
        Log.d(TAG, "loadTagTitles() called with: recipe = [" + recipe + "]");
        final String tags = recipe.getTag();
        if (TextUtils.isEmpty(tags)) return null;
        final String[] split = tags.split(",");
        return Arrays.asList(split);
    }

    @Override
    public void updateRecipe(Recipe recipe) {
        Log.d(TAG, "deleteFavorite() called with: recipe = [" + recipe + "]");
        mDetailRepository.updateRecipe(recipe);
    }

    @Override
    public void insertToFavorite(Recipe recipe) {
        Log.d(TAG, "insertToFavorite() called with: recipe = [" + recipe + "]");
        mDetailRepository.insertFavorite(recipe);
    }

    @Override
    public void removeFromFavorite(Recipe recipe) {
        Log.d(TAG, "removeFromFavorite() called with: recipe = [" + recipe + "]");
        mDetailRepository.removeFavorite(recipe);
    }

    @Override
    public Recipe loadFavorite(Recipe recipe) {
        Log.d(TAG, "loadFavorite() called with: recipe = [" + recipe + "]");
        return mDetailRepository.getFavoriteByRecipeId(recipe);
    }

    @Override
    public String[] loadTags(Recipe recipe) {
        Log.d(TAG, "loadTags() called with: recipe = [" + recipe + "]");
        final String tags = recipe.getTag();
        if (TextUtils.isEmpty(tags)) return null;
        final String[] splitTags = tags.split(",");
        return splitTags;
    }

    @Override
    public void searchByTag(String tag) {
        Log.d(TAG, "searchByTag() called with: tag = [" + tag + "]");
        mDetailView.navigateToSearchWithTag(tag);
    }
}

File: app/src/main/java/com/zest/android/detail/DetailPresenter.kt
package com.zest.android.detail

import android.text.TextUtils
import android.util.Log
import com.zest.android.data.Recipe
import com.zest.android.data.source.DetailRepository


/**
 * Listens to user actions from [DetailActivity], retrieves the data and updates the
 * UI as required.
 *
 * Created by ZARA on 09/25/2018.
 */
class DetailPresenter(private val detailView: DetailContract.View,
                      private val detailRepository: DetailRepository) : DetailContract.UserActionsListener {

    init {
        detailView.setPresenter(this)
    }

    override fun start() {
    }

    override fun updateRecipe(recipe: Recipe) {
        Log.d(TAG, "deleteFavorite() called with: recipe = [$recipe]")
        detailRepository.updateRecipe(recipe)
    }

    override fun insertToFavorite(recipe: Recipe) {
        Log.d(TAG, "insertToFavorite() called with: recipe = [$recipe]")
        detailRepository.insertFavorite(recipe)
    }

    override fun removeFromFavorite(recipe: Recipe) {
        Log.d(TAG, "removeFromFavorite() called with: recipe = [$recipe]")
        detailRepository.removeFavorite(recipe)
    }

    override fun loadFavorite(recipe: Recipe): Recipe? {
        Log.d(TAG, "loadFavorite() called with: recipe = [$recipe]")
        return detailRepository.getFavoriteByRecipeId(recipe)
    }

    override fun loadTags(recipe: Recipe): Array<String>? {
        Log.d(TAG, "loadTags() called with: recipe = [$recipe]")
        val tags = recipe.tag
        if (TextUtils.isEmpty(tags)) return null
        val splitTags = tags?.split(",".toRegex())?.dropLastWhile({ it.isEmpty() })?.toTypedArray()
        return splitTags
    }

    override fun searchByTag(tag: String) {
        Log.d(TAG, "searchByTag() called with: tag = [$tag]")
        detailView.navigateToSearchWithTag(tag)
    }

    companion object {
        private val TAG = DetailPresenter::class.java.simpleName
    }
}